In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
import os
print(os.getcwd())

C:\Users\Dell\OneDrive\Desktop\RetailHub-India-Business-Intelligence\05_Python


In [3]:
import os

print(os.listdir(".."))

['.git', '.gitignore', '01_business_requirements', '02_raw_data', '03_cleaned_data', '04_SQL', '05_Python', '06_Power_BI', '07_report', '08_Presentation', '09_Images', '10_Dataset_dictionary', 'README.md']


In [4]:
customer=pd.read_csv("../02_raw_data/customers.csv")
order_items=pd.read_csv("../02_raw_data/order_line_items.csv")
orders=pd.read_csv("../02_raw_data/orders.csv")
catalog=pd.read_csv("../02_raw_data/sku_catalog.csv")
sessions=pd.read_csv("../02_raw_data/website_sessions.csv")

## Data Inventory ##
In this step, we inspect the size of each dataset to understand the amount of data available for analysis.

In [5]:
customer.shape

(16865, 11)

In [6]:
orders.shape

(30000, 16)

In [7]:
order_items.shape

(42121, 10)

In [8]:
catalog.shape

(55, 5)

In [9]:
sessions.shape

(475658, 14)

## Understanding the Columns

In this step, we inspect the columns of each table to understand what business information each dataset contains.

In [10]:
customer.columns

Index(['Customer ID', 'Name', 'First order date', 'Total orders',
       'Total revenue', 'Average order value', 'Time to 2nd purchase',
       'Last purchase date', 'City / tier',
       'Acquisition channel (first touch)', 'RePurchased'],
      dtype='object')

In [11]:
customer.sample(10)

,Customer ID,Name,First order date,Total orders,Total revenue,Average order value,Time to 2nd purchase,Last purchase date,City / tier,Acquisition channel (first touch),RePurchased
6110,CUST-0009807,Nidhi D’Alia,06-07-2024,5,14108,2821.60,297.0,25-09-2025,Mumbai / Tier 1,Meta,Y
1253,CUST-0002052,Ikshita Banerjee,22-11-2024,2,6169,3084.50,55.0,17-01-2025,Bengaluru / Tier 1,Email/SMS,Y
12478,CUST-0019961,Mitali Pall,24-11-2025,1,2531,2531.00,NaN,24-11-2025,Indore / Tier 2,Influencer,N
16803,CUST-0026900,Ekaraj Natt,12-05-2023,4,9901,2475.25,14.0,01-03-2025,Gurgaon / Tier 2,Meta,Y
16421,CUST-0026263,Nitara Ray,01-11-2023,3,7571,2523.67,237.0,28-10-2025,Bhubaneswar / Tier 2,Google,Y
12515,CUST-0020014,Dominic Dara,14-09-2022,2,4944,2472.00,1110.0,28-09-2025,Ludhiana / Tier 2,Email/SMS,Y
16221,CUST-0025954,Garima Yohannan,08-10-2022,2,4606,2303.00,1159.0,11-12-2025,Hyderabad / Tier 1,Google,Y
13598,CUST-0021694,Qushi Swaminathan,08-01-2023,1,2081,2081.00,NaN,08-01-2023,Patna / Tier 2,Google,N
5987,CUST-0009622,Lila Shere,24-11-2022,1,1983,1983.00,NaN,24-11-2022,Chennai / Tier 1,Google,N
7051,CUST-0011314,Yagnesh Mukhopadhyay,17-08-2025,1,3244,3244.00,NaN,17-08-2025,Mumbai / Tier 1,Influencer,N


In [12]:
orders.columns

Index(['Order ID', 'Customer ID', 'Order date & time', 'Product',
       'Order value (gross, net)', 'Order value (gross)', 'Order value (net)',
       'Discount applied (₹ + %)', 'Discount applied (₹)',
       'Discount applied (%)', 'Payment mode', 'Shipping city', 'Pincode',
       'First order vs repeat', 'Channel source (last touch)',
       'Delivered / Returned / RTO'],
      dtype='object')

In [13]:
orders.sample(5)

,Order ID,Customer ID,Order date & time,Product,"Order value (gross, net)",Order value (gross),Order value (net),Discount applied (₹ + %),Discount applied (₹),Discount applied (%),Payment mode,Shipping city,Pincode,First order vs repeat,Channel source (last touch),Delivered / Returned / RTO
26617,ORD-20250817-026618,CUST-0012904,17-08-2025 14:15,Single item,"3,68,83,175",3688,3175,"513,13.9%",513,13.90%,COD,Pune,391917,Repeat,Influencer,Delivered
10962,ORD-20230906-010963,CUST-0019752,06-09-2023 16:53,Single item,"2,52,02,217",2520,2217,"303,12.0%",303,12.00%,Wallet,Patna,428315,Repeat,Influencer,Delivered
2740,ORD-20220627-002741,CUST-0004372,27-06-2022 16:39,2-item cart,"2,14,51,999",2145,1999,"146,6.8%",146,6.80%,COD,Chennai,378110,First,Email/SMS,Delivered
21474,ORD-20241220-021475,CUST-0007672,20-12-2024 19:09,Mixed Cart,"4,42,93,212",4429,3212,"1217,27.5%",1217,27.50%,COD,Patna,655069,First,Meta,Delivered
19589,ORD-20241001-019590,CUST-0016560,01-10-2024 17:36,Mixed Cart,"2,97,52,676",2975,2676,"298,10.0%",298,10.00%,COD,Bengaluru,324553,Repeat,Meta,Delivered


In [14]:
order_items.columns

Index(['Order ID', 'SKU ID', 'Category (top, bottom, outerwear, etc.)', 'Size',
       'Color', 'MRP', 'Selling price', 'Discount %', 'Returned? (Y/N)',
       'Return reason (if any)'],
      dtype='object')

In [15]:
order_items.sample(5)

,Order ID,SKU ID,"Category (top, bottom, outerwear, etc.)",Size,Color,MRP,Selling price,Discount %,Returned? (Y/N),Return reason (if any)
19445,ORD-20240101-013851,SKU-SHO-0033,shorts,M,White,2099,2676,0.0,N,NaN
8040,ORD-20230104-005787,SKU-PAN-0002,pants,XXL,Blue,2699,2748,0.0,N,NaN
1255,ORD-20220228-000898,SKU-PAN-0018,pants,M,Black,2799,999,50.0,Y,Different than images
4602,ORD-20220804-003277,SKU-SHI-0006,shirts,M,Grey,2799,999,50.0,N,NaN
35358,ORD-20250531-025231,SKU-PAN-0002,pants,L,Beige,2699,1220,50.0,Y,RTO


In [16]:
catalog.columns

Index(['SKU', 'Category', 'Vendor', 'MRP', 'Cost_per_unit'], dtype='object')

In [17]:
catalog.head(10)

,SKU,Category,Vendor,MRP,Cost_per_unit
0,SKU-SHI-0001,shirts,UrbanWeave Co.,1999,705
1,SKU-PAN-0002,pants,Apex Fashions,2699,982
2,SKU-SHI-0003,shirts,KnitCraft Studio,2799,1004
3,SKU-TSH-0004,tshirts,StitchWorks India,1499,539
4,SKU-CRO-0005,croptops,StitchWorks India,2399,846
5,SKU-SHI-0006,shirts,KnitCraft Studio,2799,1097
6,SKU-CRO-0007,croptops,Apex Fashions,1899,818
7,SKU-PAN-0008,pants,StitchWorks India,2999,1152
8,SKU-PAN-0009,pants,KnitCraft Studio,2799,1008
9,SKU-TSH-0010,tshirts,WestCo Garments,1799,782


In [18]:
sessions.columns

Index(['session_id', 'date', 'traffic_source', 'campaign_name',
       'device_category', 'city', 'sessions', 'product_views', 'add_to_cart',
       'begin_checkout', 'purchased', 'order_id', 'customer_id', 'revenue'],
      dtype='object')

In [19]:
sessions.sample(10)

,session_id,date,traffic_source,campaign_name,device_category,city,sessions,product_views,add_to_cart,begin_checkout,purchased,order_id,customer_id,revenue
353536,SESS-20250109-ME-MO-885523,09-01-2025,Meta,KS_AlwaysOn,mobile,Bengaluru,1,2,0,0,0,NaN,NaN,0
133188,SESS-20230406-ME-MO-170338,06-04-2023,Meta,KS_Apr23_Camp_10,mobile,Hyderabad,1,4,0,0,0,NaN,NaN,0
340945,SESS-20241205-ME-DE-604660,05-12-2024,Meta,KS_Dec24_Camp_34,desktop,Ahmedabad,1,3,1,1,0,NaN,NaN,0
193614,SESS-20231010-ME-MO-242693,10-10-2023,Meta,KS_Oct23_Camp_20,mobile,Jaipur,1,1,0,0,0,NaN,NaN,0
131983,SESS-20230402-IN-MO-125404,02-04-2023,Influencer,NaN,mobile,Jaipur,1,5,0,0,0,NaN,NaN,0
438227,SESS-20250925-GO-MO-455989,25-09-2025,Google,NaN,mobile,Surat,1,7,0,0,0,NaN,NaN,0
468625,SESS-20251212-ME-MO-957574,12-12-2025,Meta,KS_AlwaysOn,mobile,Bengaluru,1,2,0,0,0,NaN,NaN,0
211388,SESS-20231121-IN-MO-485920,21-11-2023,Influencer,NaN,mobile,Mumbai,1,1,0,0,0,NaN,NaN,0
299841,SESS-20240815-ME-MO-178961,15-08-2024,Meta,KS_Aug24_Camp_29,mobile,Hyderabad,1,3,0,0,0,NaN,NaN,0
115551,SESS-20230213-GO-DE-804508,13-02-2023,Google,NaN,desktop,Lucknow,1,2,0,0,0,NaN,NaN,0


## Data Types

In this step, we inspect the data types of each column to identify whether they are stored correctly for analysis.

In [100]:
customer.info()
orders.info()
order_items.info()
catalog.info()
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16865 entries, 0 to 16864
Data columns (total 12 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Customer ID                        16865 non-null  object        
 1   Name                               16865 non-null  object        
 2   First order date                   16865 non-null  datetime64[ns]
 3   Total orders                       16865 non-null  int64         
 4   Total revenue                      16865 non-null  int64         
 5   Average order value                16865 non-null  float64       
 6   Time to 2nd purchase               7748 non-null   float64       
 7   Last purchase date                 16865 non-null  datetime64[ns]
 8   Acquisition channel (first touch)  16865 non-null  object        
 9   RePurchased                        16865 non-null  object        
 10  city                              

## Missing Value Analysis

In this step, we identify missing values in each table to understand data completeness and determine whether the missing values are valid business cases or data quality issues.


In [21]:
customer.isnull().sum().sort_values(ascending=False)

Time to 2nd purchase                 9117
Customer ID                             0
Name                                    0
First order date                        0
Total orders                            0
Total revenue                           0
Average order value                     0
Last purchase date                      0
City / tier                             0
Acquisition channel (first touch)       0
RePurchased                             0
dtype: int64

In [22]:
orders.isnull().sum().sort_values(ascending=False)

Order ID                       0
Customer ID                    0
Order date & time              0
Product                        0
Order value (gross, net)       0
Order value (gross)            0
Order value (net)              0
Discount applied (₹ + %)       0
Discount applied (₹)           0
Discount applied (%)           0
Payment mode                   0
Shipping city                  0
Pincode                        0
First order vs repeat          0
Channel source (last touch)    0
Delivered / Returned / RTO     0
dtype: int64

In [23]:
order_items.isnull().sum().sort_values(ascending=False)

Return reason (if any)                     27244
Order ID                                       0
SKU ID                                         0
Category (top, bottom, outerwear, etc.)        0
Size                                           0
Color                                          0
MRP                                            0
Selling price                                  0
Discount %                                     0
Returned? (Y/N)                                0
dtype: int64

In [24]:
catalog.isnull().sum().sort_values(ascending=False)

SKU              0
Category         0
Vendor           0
MRP              0
Cost_per_unit    0
dtype: int64

In [25]:
sessions.isnull().sum().sort_values(ascending=False)

order_id           450894
customer_id        450894
campaign_name      247466
session_id              0
date                    0
traffic_source          0
device_category         0
city                    0
sessions                0
product_views           0
add_to_cart             0
begin_checkout          0
purchased               0
revenue                 0
dtype: int64

## Data type fix 
in this step, we will fix the dates column to data type date and other numeric values to float or int 

In [26]:
customer['First order date']=pd.to_datetime(customer['First order date'],dayfirst="True")

In [27]:
orders["Order date & time"]=pd.to_datetime(orders["Order date & time"],dayfirst="True")

In [28]:
orders["Discount applied (%)"]=(orders["Discount applied (%)"].str.replace("%","").astype(float))

In [29]:
customer['Last purchase date']=pd.to_datetime(customer['Last purchase date'],dayfirst="True")

In [30]:
sessions["date"]=pd.to_datetime(sessions["date"],dayfirst="True")

In [31]:
customer.info()
orders.info()
order_items.info()
catalog.info()
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16865 entries, 0 to 16864
Data columns (total 11 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Customer ID                        16865 non-null  object        
 1   Name                               16865 non-null  object        
 2   First order date                   16865 non-null  datetime64[ns]
 3   Total orders                       16865 non-null  int64         
 4   Total revenue                      16865 non-null  int64         
 5   Average order value                16865 non-null  float64       
 6   Time to 2nd purchase               7748 non-null   float64       
 7   Last purchase date                 16865 non-null  datetime64[ns]
 8   City / tier                        16865 non-null  object        
 9   Acquisition channel (first touch)  16865 non-null  object        
 10  RePurchased                       

## Duplicate Check

Duplicate records can lead to incorrect analysis by inflating sales, customers, or orders.

In this step, we will identify duplicate rows in each table and decide whether they should be removed.

In [32]:
print(customer.duplicated().sum())

0


In [33]:
print(orders.duplicated().sum())

0


In [34]:
print(order_items.duplicated().sum())

13


In [35]:
print(catalog.duplicated().sum())

0


In [36]:
print(sessions.duplicated().sum())

2


## removing duplicates

In [37]:
order_items[order_items.duplicated(keep=False)]

,Order ID,SKU ID,"Category (top, bottom, outerwear, etc.)",Size,Color,MRP,Selling price,Discount %,Returned? (Y/N),Return reason (if any)
1307,ORD-20220303-000934,SKU-SHI-0001,shirts,L,Navy,1999,999,50.0,N,NaN
1308,ORD-20220303-000934,SKU-SHI-0001,shirts,L,Navy,1999,999,50.0,N,NaN
5844,ORD-20221006-004172,SKU-SHI-0001,shirts,XXL,Black,1999,999,50.0,N,NaN
5845,ORD-20221006-004172,SKU-SHI-0001,shirts,XXL,Black,1999,999,50.0,N,NaN
6933,ORD-20221115-004977,SKU-SHI-0001,shirts,XS,Grey,1999,999,50.0,N,NaN
6934,ORD-20221115-004977,SKU-SHI-0001,shirts,XS,Grey,1999,999,50.0,N,NaN
11713,ORD-20230429-008398,SKU-SHI-0003,shirts,M,Black,2799,999,50.0,N,NaN
11714,ORD-20230429-008398,SKU-SHI-0003,shirts,M,Black,2799,999,50.0,N,NaN
12043,ORD-20230513-008624,SKU-PAN-0002,pants,M,Olive,2699,999,50.0,N,NaN
12044,ORD-20230513-008624,SKU-PAN-0002,pants,M,Olive,2699,999,50.0,N,NaN


In [38]:
order_items.drop_duplicates(inplace=True)

In [39]:
sessions[sessions.duplicated(keep=False)]

,session_id,date,traffic_source,campaign_name,device_category,city,sessions,product_views,add_to_cart,begin_checkout,purchased,order_id,customer_id,revenue
44733,SESS-20220615-ME-MO-755749,2022-06-15,Meta,KS_AlwaysOn,mobile,Pune,1,4,0,0,0,NaN,NaN,0
44760,SESS-20220615-ME-MO-755749,2022-06-15,Meta,KS_AlwaysOn,mobile,Pune,1,4,0,0,0,NaN,NaN,0
322983,SESS-20241019-ME-MO-409533,2024-10-19,Meta,KS_Oct24_Camp_31,mobile,Surat,1,1,0,0,0,NaN,NaN,0
323005,SESS-20241019-ME-MO-409533,2024-10-19,Meta,KS_Oct24_Camp_31,mobile,Surat,1,1,0,0,0,NaN,NaN,0


In [40]:
sessions.drop_duplicates(inplace=True)

## Removing unwanted columns

In [41]:
orders.drop("Order value (gross, net)",axis=1,inplace=True)

In [42]:
orders.drop("Discount applied (₹ + %)",axis=1,inplace=True)

In [43]:
orders.columns

Index(['Order ID', 'Customer ID', 'Order date & time', 'Product',
       'Order value (gross)', 'Order value (net)', 'Discount applied (₹)',
       'Discount applied (%)', 'Payment mode', 'Shipping city', 'Pincode',
       'First order vs repeat', 'Channel source (last touch)',
       'Delivered / Returned / RTO'],
      dtype='object')

## Spliting columns 

In [44]:
customer[["city","tier"]]=customer["City / tier"].str.split("/",expand=True)

In [45]:
customer.drop("City / tier",axis=1,inplace=True)

In [46]:
customer.columns

Index(['Customer ID', 'Name', 'First order date', 'Total orders',
       'Total revenue', 'Average order value', 'Time to 2nd purchase',
       'Last purchase date', 'Acquisition channel (first touch)',
       'RePurchased', 'city', 'tier'],
      dtype='object')

In [47]:
customer[["city","tier"]].tail()

,city,tier
16860,Pune,Tier 1
16861,Hyderabad,Tier 1
16862,Vadodara,Tier 2
16863,Bhopal,Tier 2
16864,Delhi,Tier 1


## Investigating the purchased vs orders difference 

In [48]:
purchased_flagged = (sessions["purchased"] == 1).sum()

In [49]:
print("purchased = 1:", purchased_flagged)

purchased = 1: 29983


In [50]:
has_order_id=sessions["order_id"].notnull().sum()

In [51]:
print("order_id_not_null",has_order_id)

order_id_not_null 24764


In [52]:
mismatch = sessions[(sessions["purchased"] == 1) & (sessions["order_id"].isnull())]
print("Rows with purchased=1 but no order_id:", len(mismatch))
print("Revenue on these rows (should be 0 if truly not purchased):", mismatch["revenue"].sum())
mismatch[["session_id", "traffic_source", "purchased", "order_id", "revenue"]].head()

Rows with purchased=1 but no order_id: 5219
Revenue on these rows (should be 0 if truly not purchased): 0


,session_id,traffic_source,purchased,order_id,revenue
241,SESS-20220101-OR-MO-891977,Organic Instagram,1,NaN,0
242,SESS-20220101-OR-MO-386930,Organic Instagram,1,NaN,0
243,SESS-20220101-OR-MO-170254,Organic Instagram,1,NaN,0
257,SESS-20220101-OR-DE-809468,Organic Instagram,1,NaN,0
535,SESS-20220102-OR-MO-195094,Organic Instagram,1,NaN,0


In [53]:
sessions["is_converted"] = sessions["order_id"].notnull()

In [54]:
sessions["is_converted"].sum()  # should equal 24,764 (the order_id.notnull() count from before)

np.int64(24764)

In [55]:
sessions["purchased"].sum()      # still 29,983, untouched, for comparison

np.int64(29983)

**Finding:** All 5,219 mismatched rows have `revenue = 0`, meaning they don't distort revenue totals. This looks like a tracking gap — the session was flagged as "purchased" (e.g. reached a thank-you page) but the order wasn't successfully linked back (possibly a failed payment retry, or a tracking bug). 

## Incosistency check
In this step we will check the typo errors in the columns for alphabetical corrections and 

In [56]:
customer["Acquisition channel (first touch)"].unique()

array(['Meta', 'Google', 'Offline (Friends/Family)', 'Influencer',
       'Organic Instagram', 'Instagram DM', 'Email/SMS'], dtype=object)

In [57]:
customer["city"].unique()

array(['Mumbai ', 'Bengaluru ', 'Hyderabad ', 'Delhi ', 'Nashik ',
       'Noida ', 'Surat ', 'Chennai ', 'Rajkot ', 'Ahmedabad ', 'Patna ',
       'Lucknow ', 'Dehradun ', 'Indore ', 'Pune ', 'Kochi ', 'Kanpur ',
       'Coimbatore ', 'Nagpur ', 'Ludhiana ', 'Jaipur ', 'Kolkata ',
       'Mysuru ', 'Bhopal ', 'Gurgaon ', 'Bhubaneswar ', 'Visakhapatnam ',
       'Vadodara ', 'Ranchi ', 'Chandigarh '], dtype=object)

In [58]:
order_items["Color"].unique()

array(['Maroon', 'Olive', 'Black', 'Navy', 'Beige', 'Grey', 'White',
       'Blue'], dtype=object)

In [59]:
order_items["Category (top, bottom, outerwear, etc.)"].unique()

array(['pants', 'shirts', 'croptops', 'tshirts', 'shorts', 'jackets'],
      dtype=object)

In [60]:
orders["Payment mode"].unique()

array(['Card', 'UPI', 'COD', 'Wallet', 'NetBanking'], dtype=object)

In [61]:
orders["Shipping city"].unique()

array(['Lucknow', 'Delhi', 'Chandigarh', 'Indore', 'Mumbai',
       'Visakhapatnam', 'Dehradun', 'Kanpur', 'Surat', 'Bengaluru',
       'Pune', 'Bhubaneswar', 'Mysuru', 'Nagpur', 'Nashik', 'Kochi',
       'Patna', 'Ahmedabad', 'Rajkot', 'Bhopal', 'Hyderabad', 'Ranchi',
       'Jaipur', 'Noida', 'Chennai', 'Ludhiana', 'Vadodara', 'Coimbatore',
       'Kolkata', 'Gurgaon'], dtype=object)

In [62]:
orders["First order vs repeat"].unique()

array(['First', 'Repeat'], dtype=object)

In [63]:
orders["Channel source (last touch)"].unique()

array(['Meta', 'Organic Instagram', 'Influencer', 'Google', 'Email/SMS',
       'Offline (Friends/Family)', 'Instagram DM'], dtype=object)

In [64]:
orders["Delivered / Returned / RTO"].unique()

array(['Delivered', 'RTO', 'Returned'], dtype=object)

In [65]:
catalog["Category"].unique()

array(['shirts', 'pants', 'tshirts', 'croptops', 'shorts', 'jackets'],
      dtype=object)

In [66]:
catalog["Vendor"].unique()

array(['UrbanWeave Co.', 'Apex Fashions', 'KnitCraft Studio',
       'StitchWorks India', 'WestCo Garments'], dtype=object)

sessions["traffic_source"].unique()

In [67]:
sessions["campaign_name"].unique()

array(['KS_AlwaysOn', nan, 'KS_Jan23_Camp_02', 'KS_Jan23_Camp_01',
       'KS_Mar23_Camp_06', 'KS_Mar23_Camp_09', 'KS_Mar23_Camp_05',
       'KS_Mar23_Camp_04', 'KS_Mar23_Camp_08', 'KS_Mar23_Camp_07',
       'KS_Mar23_Camp_03', 'KS_Apr23_Camp_11', 'KS_Apr23_Camp_12',
       'KS_Apr23_Camp_10', 'KS_May23_Camp_13', 'KS_Jul23_Camp_14',
       'KS_Aug23_Camp_15', 'KS_Sep23_Camp_17', 'KS_Sep23_Camp_18',
       'KS_Sep23_Camp_16', 'KS_Oct23_Camp_19', 'KS_Oct23_Camp_20',
       'KS_Nov23_Camp_21', 'KS_Dec23_Camp_22', 'KS_Mar24_Camp_23',
       'KS_Apr24_Camp_24', 'KS_Jun24_Camp_25', 'KS_Jul24_Camp_27',
       'KS_Jul24_Camp_28', 'KS_Jul24_Camp_26', 'KS_Aug24_Camp_29',
       'KS_Sep24_Camp_30', 'KS_Oct24_Camp_32', 'KS_Oct24_Camp_31',
       'KS_Dec24_Camp_34', 'KS_Dec24_Camp_33', 'KS_Feb25_Camp_35',
       'KS_Mar25_Camp_36', 'KS_Mar25_Camp_37', 'KS_Mar25_Camp_38',
       'KS_Apr25_Camp_39', 'KS_May25_Camp_40', 'KS_Jun25_Camp_42',
       'KS_Jun25_Camp_41', 'KS_Jul25_Camp_43', 'KS_Sep25_Camp_

In [68]:
sessions["device_category"].unique()

array(['mobile', 'desktop', 'tablet'], dtype=object)

In [69]:
sessions["city"].unique()

array(['Pune', 'Mumbai', 'Bengaluru', 'Ahmedabad', 'Jaipur', 'Surat',
       'Delhi', 'Hyderabad', 'Lucknow', 'Indore'], dtype=object)

## Business Validation
Do the numbers and relationships make business sense?

In [70]:
order_items[order_items["Selling price"] > order_items["MRP"]]

,Order ID,SKU ID,"Category (top, bottom, outerwear, etc.)",Size,Color,MRP,Selling price,Discount %,Returned? (Y/N),Return reason (if any)
2,ORD-20220101-000003,SKU-CRO-0019,croptops,L,Olive,1799,2210,0.0,N,NaN
50,ORD-20220102-000036,SKU-TSH-0004,tshirts,XL,Blue,1499,1942,0.0,Y,Different than images
61,ORD-20220103-000045,SKU-SHI-0001,shirts,XXL,White,1999,2269,0.0,N,NaN
62,ORD-20220103-000046,SKU-TSH-0004,tshirts,XXL,Olive,1499,1613,0.0,N,NaN
76,ORD-20220104-000058,SKU-CRO-0014,croptops,XS,White,1799,2257,0.0,Y,Size issue
...,...,...,...,...,...,...,...,...,...,...
42113,ORD-20251231-029995,SKU-SHI-0003,shirts,M,Black,2799,3803,0.0,N,NaN
42114,ORD-20251231-029996,SKU-SHI-0001,shirts,XS,Blue,1999,2701,0.0,N,NaN
42115,ORD-20251231-029997,SKU-PAN-0020,pants,S,White,2999,3335,0.0,N,NaN
42116,ORD-20251231-029998,SKU-SHI-0001,shirts,XXL,Grey,1999,3164,0.0,N,NaN


In [71]:
order_items[
    order_items["Selling price"] > order_items["MRP"]
][["SKU ID", "MRP", "Selling price", "Discount %"]].head(20)

,SKU ID,MRP,Selling price,Discount %
2,SKU-CRO-0019,1799,2210,0.0
50,SKU-TSH-0004,1499,1942,0.0
61,SKU-SHI-0001,1999,2269,0.0
62,SKU-TSH-0004,1499,1613,0.0
76,SKU-CRO-0014,1799,2257,0.0
90,SKU-SHI-0001,1999,2017,0.0
104,SKU-TSH-0049,1499,1960,0.0
114,SKU-TSH-0004,1499,1581,0.0
121,SKU-TSH-0041,1499,1748,0.0
134,SKU-TSH-0004,1499,1760,0.0


In [72]:
order_items[
    order_items["Selling price"] > order_items["MRP"]
][["MRP", "Selling price", "Discount %"]].describe()

,MRP,Selling price,Discount %
count,12019.000000,12019.000000,12019.0
mean,2101.712372,2829.214993,0.0
std,409.795558,530.985782,0.0
min,1499.000000,1502.000000,0.0
25%,1799.000000,2438.000000,0.0
50%,1999.000000,2839.000000,0.0
75%,2399.000000,3205.000000,0.0
max,3999.000000,4238.000000,0.0


In [73]:
order_items[["SKU ID", "MRP", "Selling price"]].head(20)

,SKU ID,MRP,Selling price
0,SKU-PAN-0002,2699,1814
1,SKU-SHI-0001,1999,1812
2,SKU-CRO-0019,1799,2210
3,SKU-SHI-0003,2799,1395
4,SKU-SHI-0045,2299,999
5,SKU-SHI-0036,2299,999
6,SKU-PAN-0002,2699,999
7,SKU-SHI-0001,1999,999
8,SKU-PAN-0009,2799,1530
9,SKU-SHI-0001,1999,1872


In [74]:
catalog[["SKU","MRP"]]

,SKU,MRP
0,SKU-SHI-0001,1999
1,SKU-PAN-0002,2699
2,SKU-SHI-0003,2799
3,SKU-TSH-0004,1499
4,SKU-CRO-0005,2399
5,SKU-SHI-0006,2799
6,SKU-CRO-0007,1899
7,SKU-PAN-0008,2999
8,SKU-PAN-0009,2799
9,SKU-TSH-0010,1799


In [76]:
order_items['Actual discount %'] = (order_items['MRP'] - order_items['Selling price']) / order_items['MRP'] * 100 
#added a new column for actual discount 

In [77]:
order_items["Actual discount %"].describe()

count    42108.000000
mean        17.407107
std         41.205048
min       -179.853235
25%         -6.307996
50%         30.605530
75%         50.025013
max         79.183163
Name: Actual discount %, dtype: float64

## Finding: `Discount %` unreliable for 12,019 rows (Selling Price > MRP)
These rows show a markup (selling price above MRP), but `Discount %` is floored at 0 instead of negative. Kept `Selling price`/`MRP` as-is, added `Actual discount %` with the correct formula for analysis, and did not drop these rows (~28% of line items, likely real demand-based pricing).

In [78]:
customer_revenue = orders.groupby("Customer ID")["Order value (net)"].sum()
customer_revenue

Customer ID
CUST-0000001     3356
CUST-0000003     2360
CUST-0000004     4029
CUST-0000005     2581
CUST-0000007     5365
                ...  
CUST-0026992     9179
CUST-0026996     7064
CUST-0026997    10087
CUST-0026999     1606
CUST-0027000     2799
Name: Order value (net), Length: 16865, dtype: int64

In [79]:
comparison = customer[["Customer ID", "Total revenue"]].merge(
    customer_revenue,
    on="Customer ID",
    how="left"
)

In [80]:
comparison["Difference"] = (
    comparison["Total revenue"] -
    comparison["Order value (net)"]
)

In [81]:
comparison["Difference"].describe()

count    16865.0
mean         0.0
std          0.0
min          0.0
25%          0.0
50%          0.0
75%          0.0
max          0.0
Name: Difference, dtype: float64

In [82]:
catalog[catalog['Cost_per_unit'] >= catalog['MRP']]

,SKU,Category,Vendor,MRP,Cost_per_unit


In [83]:
order_counts = orders.groupby("Customer ID")["First order vs repeat"].value_counts().unstack(fill_value=0)

In [84]:
check = customer[["Customer ID", "Total orders", "RePurchased"]].merge(
    order_counts,
    on="Customer ID",
    how="left"
)

In [85]:
check["orders_sum"] = check["First"] + check["Repeat"]

# Contradiction A: total orders in customers.csv doesn't match count of orders in orders.csv
contradiction_a = check[check["Total orders"] != check["orders_sum"]]

# Contradiction B: RePurchased = 'Y' but no orders marked 'Repeat'
contradiction_b = check[(check["RePurchased"] == "Y") & (check["Repeat"] == 0)]

# Contradiction C: Total orders > 1 but zero orders marked 'Repeat'
contradiction_c = check[(check["Total orders"] > 1) & (check["Repeat"] == 0)]

# Contradiction D: Total orders > 1 but RePurchased is No'
contradiction_d = check[(check["Total orders"] > 1) & (check["RePurchased"] == "N")]

print("Contradiction A (order count mismatch):", len(contradiction_a))
print("Contradiction B (RePurchased=Y but no Repeat orders):", len(contradiction_b))
print("Contradiction C (>1 order but no Repeat orders):", len(contradiction_c))
print("Total orders > 1 but RePurchased is No:",len(contradiction_d))

Contradiction A (order count mismatch): 0
Contradiction B (RePurchased=Y but no Repeat orders): 18
Contradiction C (>1 order but no Repeat orders): 18
Total orders > 1 but RePurchased is No: 0


In [86]:
set(contradiction_b["Customer ID"]) == set(contradiction_c["Customer ID"])

True

In [87]:
suspect_ids = contradiction_b["Customer ID"].tolist()
orders[orders["Customer ID"].isin(suspect_ids)].sort_values(["Customer ID", "Order date & time"])

,Order ID,Customer ID,Order date & time,Product,Order value (gross),Order value (net),Discount applied (₹),Discount applied (%),Payment mode,Shipping city,Pincode,First order vs repeat,Channel source (last touch),Delivered / Returned / RTO
794,ORD-20220220-000795,CUST-0000184,2022-02-20 16:11:00,2-item cart,2247,1933,313,13.9,UPI,Delhi,208407,First,Meta,RTO
804,ORD-20220221-000805,CUST-0000184,2022-02-21 13:28:00,2-item cart,2029,1776,253,12.5,Card,Bengaluru,496786,First,Instagram DM,RTO
14360,ORD-20240123-014361,CUST-0006366,2024-01-23 13:04:00,Single item,3448,2861,587,17.0,UPI,Nashik,440543,First,Instagram DM,Delivered
14385,ORD-20240124-014386,CUST-0006366,2024-01-24 21:01:00,Mixed Cart,3048,2744,305,10.0,UPI,Mumbai,593889,First,Google,Delivered
2131,ORD-20220514-002132,CUST-0009132,2022-05-14 10:52:00,2-item cart,2224,1925,300,13.5,UPI,Bengaluru,414368,First,Organic Instagram,Delivered
2149,ORD-20220515-002150,CUST-0009132,2022-05-15 20:14:00,Mixed Cart,2528,2201,327,12.9,UPI,Lucknow,739289,First,Google,RTO
25609,ORD-20250623-025610,CUST-0009676,2025-06-23 19:06:00,2-item cart,3889,3469,419,10.8,COD,Hyderabad,528223,First,Meta,Delivered
25617,ORD-20250624-025618,CUST-0009676,2025-06-24 14:03:00,Mixed Cart,3120,2586,534,17.1,UPI,Rajkot,415081,First,Meta,Returned
985,ORD-20220306-000986,CUST-0012729,2022-03-06 17:46:00,Single item,2351,1677,674,28.7,Card,Hyderabad,807695,First,Influencer,Returned
983,ORD-20220306-000984,CUST-0012729,2022-03-06 20:46:00,Single item,2025,1644,382,18.8,NetBanking,Nashik,495958,First,Meta,Returned


In [88]:
orders["order_rank"] = orders.groupby("Customer ID")["Order date & time"].rank(method="first")
orders["Corrected order type"] = orders["order_rank"].apply(lambda r: "First" if r == 1 else "Repeat")

In [89]:
mismatches = orders[orders["First order vs repeat"] != orders["Corrected order type"]]
print("Total mismatched rows across entire dataset:", len(mismatches))
print("Unique customers affected:", mismatches["Customer ID"].nunique())

Total mismatched rows across entire dataset: 30
Unique customers affected: 30


In [90]:
set(suspect_ids).issubset(set(mismatches["Customer ID"]))

True

In [91]:
mismatches.groupby("Customer ID")["Order date & time"].agg(lambda x: x.max() - x.min()).describe()

count                 30
mean     0 days 00:00:00
std      0 days 00:00:00
min      0 days 00:00:00
25%      0 days 00:00:00
50%      0 days 00:00:00
75%      0 days 00:00:00
max      0 days 00:00:00
Name: Order date & time, dtype: object

In [92]:
affected_ids = mismatches["Customer ID"].unique()
gaps = orders[orders["Customer ID"].isin(affected_ids)].groupby("Customer ID")["Order date & time"].agg(lambda x: x.max() - x.min())
gaps.describe()

count                             30
mean               259 days 18:36:56
std      400 days 07:53:28.213926196
min                  0 days 03:00:00
25%                  0 days 22:12:30
50%                  1 days 16:45:00
75%                550 days 21:19:15
max               1338 days 17:40:00
Name: Order date & time, dtype: object

In [93]:
large_gap_ids = gaps[gaps > pd.Timedelta(days=30)].index
orders[orders["Customer ID"].isin(large_gap_ids)].sort_values(["Customer ID", "Order date & time"])

,Order ID,Customer ID,Order date & time,Product,Order value (gross),Order value (net),Discount applied (₹),Discount applied (%),Payment mode,Shipping city,Pincode,First order vs repeat,Channel source (last touch),Delivered / Returned / RTO,order_rank,Corrected order type
14300,ORD-20240121-014301,CUST-0001008,2024-01-21 21:40:00,Mixed Cart,3586,2539,1047,29.2,Wallet,Jaipur,216088,First,Google,Delivered,1.0,First
14335,ORD-20240122-014336,CUST-0001008,2024-01-22 13:07:00,Single item,3031,2377,654,21.6,Card,Mumbai,833332,First,Organic Instagram,Delivered,2.0,Repeat
22922,ORD-20250225-022923,CUST-0001008,2025-02-25 21:06:00,2-item cart,3733,3333,400,10.7,UPI,Lucknow,147599,Repeat,Google,Delivered,3.0,Repeat
1323,ORD-20220326-001324,CUST-0002578,2022-03-26 11:13:00,Mixed Cart,1918,1528,390,20.3,Wallet,Ranchi,540034,First,Influencer,RTO,1.0,First
1322,ORD-20220326-001323,CUST-0002578,2022-03-26 19:26:00,Single item,2364,1933,431,18.2,Card,Mumbai,230765,First,Meta,Returned,2.0,Repeat
15762,ORD-20240327-015763,CUST-0002578,2024-03-27 18:29:00,Single item,3861,3126,735,19.0,COD,Delhi,606250,Repeat,Influencer,Returned,3.0,Repeat
17607,ORD-20240622-017608,CUST-0002578,2024-06-22 19:16:00,Single item,3069,2662,407,13.3,COD,Noida,598844,Repeat,Meta,Returned,4.0,Repeat
8605,ORD-20230512-008606,CUST-0004798,2023-05-12 22:25:00,2-item cart,2280,2043,236,10.4,UPI,Hyderabad,696677,First,Organic Instagram,Delivered,1.0,First
8624,ORD-20230513-008625,CUST-0004798,2023-05-13 11:30:00,Single item,2762,2523,239,8.7,COD,Pune,371858,First,Meta,Delivered,2.0,Repeat
27279,ORD-20250918-027280,CUST-0004798,2025-09-18 14:13:00,Single item,3021,2570,451,14.9,Card,Hyderabad,181751,Repeat,Meta,Delivered,3.0,Repeat


## Finding: `First order vs repeat` mislabeled for 30 customers (~0.18%)
Comparing the original label against a date-derived version (earliest order per customer = First, rest = Repeat) found 30 customers with at least one wrong label. Root cause unclear — roughly half show orders placed within 1-2 days of each other (likely a pipeline timing issue), the rest show gaps of months to years (possibly a reactivation-relabeling rule, unconfirmed). Given the small scale (<0.2% of customers), the original column is kept as-is for traceability, but a derived `Corrected order type` column is used for all repeat-purchase analysis instead.

In [94]:
catalog[catalog["Cost_per_unit"] >= catalog["MRP"]]

,SKU,Category,Vendor,MRP,Cost_per_unit


In [95]:
customer[["Total orders", "Average order value", "Time to 2nd purchase"]].describe()

,Total orders,Average order value,Time to 2nd purchase
count,16865.000000,16865.000000,7748.000000
mean,1.778832,2528.301471,410.219928
std,1.110665,531.041725,315.313487
min,1.000000,1350.000000,0.000000
25%,1.000000,2150.670000,152.000000
50%,1.000000,2518.000000,342.000000
75%,2.000000,2874.000000,606.000000
max,9.000000,4234.000000,1452.000000


In [101]:
customer.to_csv("../03_cleaned_data/customers_cleaned.csv", index=False)
orders.to_csv("../03_cleaned_data/orders_cleaned.csv", index=False)
order_items.to_csv("../03_cleaned_data/order_line_items_cleaned.csv", index=False)
catalog.to_csv("../03_cleaned_data/sku_catalog_cleaned.csv", index=False)
sessions.to_csv("../03_cleaned_data/website_sessions_cleaned.csv", index=False)

print("Exported files:")
import os
print(os.listdir("../03_cleaned_data"))

Exported files:
['customers_cleaned.csv', 'orders_cleaned.csv', 'order_line_items_cleaned.csv', 'sku_catalog_cleaned.csv', 'website_sessions_cleaned.csv']


In [97]:
sessions.to_csv("../03_cleaned_data/website_sessions_cleaned.csv", index=False)

In [99]:
orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 16 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   Order ID                     30000 non-null  object        
 1   Customer ID                  30000 non-null  object        
 2   Order date & time            30000 non-null  datetime64[ns]
 3   Product                      30000 non-null  object        
 4   Order value (gross)          30000 non-null  int64         
 5   Order value (net)            30000 non-null  int64         
 6   Discount applied (₹)         30000 non-null  int64         
 7   Discount applied (%)         30000 non-null  float64       
 8   Payment mode                 30000 non-null  object        
 9   Shipping city                30000 non-null  object        
 10  Pincode                      30000 non-null  int64         
 11  First order vs repeat        30000 non-nu